File: day_01/scala/src/main/scala/Day01.scala
import scala.annotation.tailrec
import scala.io.Source

object Day01 {
  def loadFile(filename: String): Array[Int] =
    Source
      .fromFile(filename)
      .getLines
      .map(_.trim)
      .map(_.toInt)
      .toArray

  def part1(values: Array[Int]): Unit = {
    println(s"Answer part 1: ${values.sum}")
  }

  def part2(values: Array[Int]): Unit = {
    // Loop array infinitely.
    val infiniteValues = Stream.continually(values.toStream).flatten

    @tailrec
    def sumUntilCycle(values: Stream[Int], sum: Int = 0, visited: Set[Int] = Set.empty): Int = {
      val updatedSum = sum + values.head
      if (visited.contains(updatedSum)) {
        return updatedSum
      }
      sumUntilCycle(values.tail, updatedSum, visited + updatedSum)
    }

    println(s"Answer part 2: ${sumUntilCycle(infiniteValues)}")
  }

  def main(args: Array[String]): Unit = {
    val values = loadFile("../input.txt")
    part1(values)
    part2(values)
  }
}

File: day_01/nim/day01.nim
import strutils
import intsets

proc read_data(filename: string): seq[int] =
  var values = newSeq[int]()
  for line in lines filename:
    values.add(strutils.parseInt(line))
  return values

proc part1(values: seq[int]) =
  var sum = 0
  for value in values:
    sum += value
  echo "Part1: ", sum

proc part2(values: seq[int]) =
  var visited: IntSet
  var v = 0
  while true:
    for _, value in values:
      v += value
      if v in visited:
        echo "Part2: ", v
        return
      else:
        visited.incl(v)

let values = read_data("../input.txt")
part1(values)
part2(values)
